lib/puppet/face/facts.rb

```diff
@@ -140,7 +138,7 @@
       options[:resolve_options] = true
       result = Puppet::Node::Facts.indirection.find(Puppet.settings[:certname], options)
 
-      result.values
+      result
```

Comment:
Unfortunately the faces code is complicated and the API is not well documented, but I think there's an easier way for your action to be responsible for rendering the facts object instead of trying to "squish" the facts object and relying on the `:console` format. 

If you modify the `show` action to return the fact **values**:

```ruby
    when_invoked do |name, args|
      Puppet.settings.preferred_run_mode = :agent
      Puppet::Node::Facts.indirection.terminus_class = :facter

      options = {}
      options[:user_query] = [name]
      options[:resolve_options] = true
      result = Puppet::Node::Facts.indirection.find(Puppet.settings[:certname], options)
      result.values
    end
```

And define a `when_rendered` block for the default `:console`:

```ruby
    when_rendering :console do |result|
      Puppet::Util::Json.dump(result, :pretty => true)
    end
```

Then you'll get pretty printed JSON:

```
$ bx puppet facts show ruby
{
  "ruby": {
    "platform": "x86_64-darwin18",
    "sitedir": "/usr/local/Cellar/rbenv/1.1.2/versions/2.7.2/lib/ruby/site_ruby/2.7.0",
    "version": "2.7.2"
  }
}
```

If you specify `render_as` explicitly on the command line, then faces will use that format instead of `:console`, so `yaml` prints as expected:

```
$ bx puppet facts show ruby --render_as yaml
---
ruby:
  platform: x86_64-darwin18
  sitedir: "/usr/local/Cellar/rbenv/1.1.2/versions/2.7.2/lib/ruby/site_ruby/2.7.0"
  version: 2.7.2
```

and `--render_as json` prints non-pretty printed JSON as before.

Reply:
Thank you very much, the solution you proposed is cleaner and easier and i didn't know that a `when_rendered ` method can be implemented on actions.